ethcore/src/engines/clique/util.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+use ethereum_types::{Address, H256};
+use lru_cache::LruCache;
+use parking_lot::RwLock;
+
+use engines::clique::{SIGNER_SIG_LENGTH, SIGNER_VANITY_LENGTH};
+use error::Error;
+use ethkey::{public_to_address, recover as ec_recover, Signature};
+use types::header::Header;
+
+/// How many recovered signature to cache in the memory.
+pub const CREATOR_CACHE_NUM: usize = 4096;
+lazy_static! {
+	/// key: header hash
+	/// value: creator address
+	static ref CREATOR_BY_HASH: RwLock<LruCache<H256, Address>> = RwLock::new(LruCache::new(CREATOR_CACHE_NUM));
+}
+
+/// Recover block creator from signature
+pub fn recover_creator(header: &Header) -> Result<Address, Error> {
+	// Initialization
+	let mut cache = CREATOR_BY_HASH.write();
+
+	if let Some(creator) = cache.get_mut(&header.hash()) {
+		return Ok(*creator);
+	}
+
+	let data = header.extra_data();
+	let mut sig_data = data[data.len() - SIGNER_SIG_LENGTH..].to_vec();
```

Comment:
How is it verified that `data.len() >= SIGNER_SIG_LENGTH`?

If the first condition holds I suggest to use [slice::split()](https://doc.rust-lang.org/std/primitive.slice.html#method.split) instead of manually splitting it you can use something like:

```rust
   let boundry = header.extra_data().len() - SIGNER_SIG_LENGTH;
   let (extra_data, sign_data) = &header.extra_data().split_at(boundry);
   // convert to vec push and so on
```

Reply:
added a explict check.